### PR TITLE
Issue #302: Implement checkNameForContextStrategyByTokenOrAncestorSet

### PR DESCRIPTION
**Pull request by HuGanghui**
Issue #302: Implement checkNameForContextStrategyByTokenOrAncestorSet

this PR update `checkNameForContextStrategyByTokenOrParentSet` to `checkNameForContextStrategyByTokenOrAncestorSet`. This update does not affect original checks listed in https://github.com/checkstyle/patch-filters/issues/302#issuecomment-674096024 that using `ParentSet`, and will make 14 checks not need to use `neverSuppressedChecks ` now, which will solve or relieve the problem is mentioned in issue #125

### CODE REVIEW

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionpatchxpathfilter/EmptyStatement/patchedline/defaultContext.patch [hunk 1]
```diff
@@ -16,3 +17,13 @@ index 8c68b53..acd378e 100644
...
+=======
+@@ -5,7 +5,6 @@ public class Test {
+         switch (i)
+         {
+             case 1 :
+-                break
+                 ;  // violation without filter
+             default :
+                 ;  // violation without filter
+>>>>>>> 77241675... Issue #302: Implement checkNameForContextStrategyByTokenOrAncestor
```
Comment (baratali) [inline src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionpatchxpathfilter/EmptyStatement/patchedline/defaultContext.patch:29]: is it part of the diff?

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/PatchXpathFilterElement.java [hunk 2]
```diff
@@ -43,11 +45,24 @@
...
+                    TokenTypes.INTERFACE_DEF,
+                    TokenTypes.CLASS_DEF
+            );
+
     /**
      * Set of checks that support context strategy but need modify violation nodes
      * to their parent abstract nodes to get their child nodes.
      */
-    private final Set<String> checkNamesForContextStrategyByTokenOrParentSet = new HashSet<>();
+    private final Set<String> checkNamesForContextStrategyByTokenOrAncestorSet = new HashSet<>();
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/PatchXpathFilterElement.java:65]: we need shorter name for this :)
let's think

**Review by baratali — CHANGES_REQUESTED** [review]

File: config/suppressions.xml [hunk 3]
```diff
@@ -15,4 +15,6 @@
             files="GeneratePatchFileLauncher|GeneratePatchFileWithGitCommandLauncher\.java"/>
     <suppress checks="UncommentedMain"
             files="GeneratePatchFileLauncher|GeneratePatchFileWithGitCommandLauncher\.java"/>
+    <suppress checks="AvoidDoubleBraceInitialization"
+              files="JavaPatchFilterElement\.java"/>
```
Comment (baratali) [inline config/suppressions.xml:19]: Do not suppress checkstyle violations. Fix it

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 4]
```diff
@@ -43,12 +45,43 @@
...
+            put("DeclarationOrder", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("EmptyStatement", Arrays.asList(TokenTypes.METHOD_DEF));
+            put("EqualsHashCode", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("FallThrough", Arrays.asList(TokenTypes.METHOD_DEF));
+            put("FinalLocalVariable", Arrays.asList(TokenTypes.METHOD_DEF,
+                    TokenTypes.VARIABLE_DEF));
+            put("InnerTypeLast", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("JavadocVariable", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("MissingOverride", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("MissingJavadocMethod", Arrays.asList(TokenTypes.CLASS_DEF));
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:67]: As discussed in https://github.com/checkstyle/patch-filters/issues/302#issuecomment-679387328, please remove these checks mapping to CLASS_DEF. Probably all CLAS_DEF mapping should be deleted, let's think

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 5]
```diff
@@ -43,12 +45,43 @@
      */
     private static final String MAX = "max";
 
+    /**
+     * Map that mapping between a check and its ancestor token types.
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:49]: `Map that mapping between a check and its ancestor token types.`
-->
`Mapping between a check and its ancestor token types.`

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 6]
```diff
@@ -43,12 +45,43 @@
...
+            put("EqualsHashCode", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("FallThrough", Arrays.asList(TokenTypes.METHOD_DEF));
+            put("FinalLocalVariable", Arrays.asList(TokenTypes.METHOD_DEF,
+                    TokenTypes.VARIABLE_DEF));
+            put("InnerTypeLast", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("JavadocVariable", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("MissingOverride", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("MissingJavadocMethod", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("RightCurly", Arrays.asList(TokenTypes.LITERAL_TRY,
+                    TokenTypes.LITERAL_IF));
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:69]: please make it one line

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 7]
```diff
@@ -245,6 +305,10 @@ private String getCheckShortName(TreeWalkerAuditEvent event) {
         return checkNames[checkNames.length - 1];
     }
 
+    private String getShortName(TreeWalkerAuditEvent event) {
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:308]: Confusing name, because we already have getCheckShortName().
Please rename:
`getCheckShortName()` -> `getCheckName()`
`getShortName()` -> `getCheckShortName()`

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 8]
```diff
@@ -231,10 +270,31 @@ private boolean isMatchingByContextStrategy(TreeWalkerAuditEvent event) {
...
+            }
+        }
+        return eventAst;
+    }
+
     private boolean containsShortName(Set<String> checkNameSet,
                                       TreeWalkerAuditEvent event) {
         final String checkShortName = getCheckShortName(event);
-        final String shortName = checkShortName.replaceAll("Check", "");
+        final String shortName = getShortName(event);
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:297]: Please rename:
`checkShortName` -> `checkName`
`shortName` -> `checkShortName`

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 9]
```diff
@@ -43,12 +45,43 @@
      */
     private static final String MAX = "max";
 
+    /**
+     * Map that mapping between a check and its ancestor token types.
+     */
+    private static final Map<String, List<Integer>> CHECK_TO_ANCESTOR_NODES_MAP =
+            new HashMap<String, List<Integer>>() {
+        {
+            put("ArrayTrailingComma", Arrays.asList(TokenTypes.METHOD_DEF));
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:54]: What if array was defined in a field or constructor?
Also, under context strategy this check should detect only changes inside an array definition, other statements in method should not trigger its violations. METHOD_DEF is a big scope for it. Please find the closest node which covers whole array in AST.

**Review by baratali — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 10]
```diff
@@ -43,12 +45,43 @@
...
+            put("AvoidNestedBlocks", Arrays.asList(TokenTypes.METHOD_DEF));
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:55]: what if nested block is not in a method?
I think SLIST node is what this Check needs to cover all cases. Please recheck.
Example AST: http://rveach.no-ip.org/checkstyle/checkstyle.php?action=view&config=72fba57f86eaf7df54f418b381c28422&code=1aff4d8fc71b843b8429b429e5cd64a8&checkstyle=checkstyle-8.35-all.jar&printTree=true&printSuppressions=

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 11]
```diff
@@ -43,12 +45,43 @@
...
+            put("CommentsIndentation", Arrays.asList(TokenTypes.METHOD_DEF));
+            put("DefaultComesLast", Arrays.asList(TokenTypes.METHOD_DEF));
+            put("DeclarationOrder", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("EmptyStatement", Arrays.asList(TokenTypes.METHOD_DEF));
+            put("EqualsHashCode", Arrays.asList(TokenTypes.CLASS_DEF));
+            put("FallThrough", Arrays.asList(TokenTypes.METHOD_DEF));
+            put("FinalLocalVariable", Arrays.asList(TokenTypes.METHOD_DEF,
+                    TokenTypes.VARIABLE_DEF));
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:63]: probably need CTOR_DEF as well?

**Review by baratali — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 6 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:69]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 5 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:49]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 7 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:308]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 8 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:297]: done

File: config/suppressions.xml (hunk 3 above)
Comment (HuGanghui) [reply config/suppressions.xml:19]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 4 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:67]: @baratali please have a look https://github.com/checkstyle/patch-filters/issues/302#issuecomment-679514926

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 9 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:54]: TokenTypes.VARIABLE_DEF is ok

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 10 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:55]: @baratali this example can not be covered if we use SLIST node, so I think we still need use METHOD_DEF

```
-        if (x == 1)
         {  // violation without filter
             x = 2;
         }
```

```
|--LITERAL_IF -> if [18:8]
    |       |   |--LPAREN -> ( [18:11]
    |       |   |--EXPR -> EXPR [18:14]
    |       |   |   `--EQUAL -> == [18:14]
    |       |   |       |--IDENT -> x [18:12]
    |       |   |       `--NUM_INT -> 1 [18:17]
    |       |   |--RPAREN -> ) [18:18]
    |       |   `--SLIST -> { [19:8]
    |       |       |--EXPR -> EXPR [20:14]
    |       |       |   `--ASSIGN -> = [20:14]
    |       |       |       |--SINGLE_LINE_COMMENT -> // [19:11]
    |       |       |       |   `--COMMENT_CONTENT ->  violation without filter\n [19:13]
    |       |       |       |--IDENT -> x [20:12]
    |       |       |       `--NUM_INT -> 2 [20:16]
    |       |       |--SEMI -> ; [20:17]
    |       |       `--RCURLY -> } [21:8]
```

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 11 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:63]: yes, I add it.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 9 above)
Comment (baratali) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java]: What if array is defined without a variable? For example:
```java
       toStream(new int[]{
         1,
         2,
         3 
       });
```
I think you need **ARRAY_INIT** node. AST example: http://rveach.no-ip.org/checkstyle/checkstyle.php?action=view&config=72fba57f86eaf7df54f418b381c28422&code=5d8cb1b4dd4dc06485c792175f91b92e&checkstyle=checkstyle-8.35-all.jar&printTree=true&printSuppressions=

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 10 above)
Comment (baratali) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java]: If your example is in method for example, then SLIST will be a method's body. If it's in other block, then SLIST will be this block's body.
Example AST with a method: http://rveach.no-ip.org/checkstyle/checkstyle.php?action=view&config=72fba57f86eaf7df54f418b381c28422&code=f44894f4f1f5e0f263488ab32aa9c3f3&checkstyle=checkstyle-8.35-all.jar&printTree=true&printSuppressions=

Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java]: yes, it's ok. changed

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 9 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java]: ok, done.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 12]
```diff
@@ -43,12 +45,55 @@
...
     private static final String MAX = "max";
 
+    /**
+     * Mapping between a check and its ancestor token types.
+     */
+    private static final Map<String, List<Integer>> CHECK_TO_ANCESTOR_NODES_MAP = new HashMap<>();
+
+    static {
+        CHECK_TO_ANCESTOR_NODES_MAP.put("ArrayTrailingComma",
+                Arrays.asList(TokenTypes.VARIABLE_DEF, TokenTypes.ARRAY_INIT));
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:55]: why do we need VARIABLE_DEF? Isn't ARRAY_INIT enough?

**Review by baratali — CHANGES_REQUESTED** [review]
Please keep only these checks in this pull request, so we can start merging faster:
 - AvoidNestedBlocks
 - FinalLocalVariable
 - RightCurly

Others are not ready yet, because we still discuss them. And it's hard to review tests for all Checks in one pull request. Please prepare one pull request for each check then.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 12 above)
Comment (HuGanghui) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:55]: removed, it is enough.

**Comment by HuGanghui** [comment]
@baratali 

>Others are not ready yet, because we still discuss them. And it's hard to review tests for all Checks in one pull request. Please prepare one pull request for each check then.

done

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/DesignForExtension/context/defaultContextConfig.xml [hunk 13]
```diff
@@ -11,8 +11,7 @@
         <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionJavaPatchFilter">
             <property name="file" value="${tp}/defaultContext.patch" />
             <property name="strategy" value="context" />
-            <property name="checkNamesForContextStrategyByTokenOrParentSet"
-                      value="DesignForExtensionCheck" />
```
Comment (baratali) [inline src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/DesignForExtension/context/defaultContextConfig.xml:15]: why is it here?

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/FallThrough/case1/context/defaultContextConfig.xml [hunk 14]
```diff
@@ -9,7 +9,7 @@
         <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionJavaPatchFilter">
             <property name="file" value="${tp}/defaultContext.patch" />
             <property name="strategy" value="context" />
-            <property name="neverSuppressedChecks" value="FallThroughCheck" />
```
Comment (baratali) [inline src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/FallThrough/case1/context/defaultContextConfig.xml:12]: how is this related to this PR?

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/FallThrough/case2/context/defaultContextConfig.xml [hunk 15]
```diff
@@ -9,7 +9,7 @@
         <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionJavaPatchFilter">
             <property name="file" value="${tp}/defaultContext.patch" />
             <property name="strategy" value="context" />
-            <property name="neverSuppressedChecks" value="FallThroughCheck" />
```
Comment (baratali) [inline src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/FallThrough/case2/context/defaultContextConfig.xml:12]: how is this related to this PR?

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/RightCurly/context/defaultContextConfig.xml [hunk 16]
```diff
@@ -9,7 +9,7 @@
         <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionJavaPatchFilter">
             <property name="file" value="${tp}/defaultContext.patch" />
             <property name="strategy" value="context" />
-            <property name="neverSuppressedChecks" value="RightCurlyCheck" />
+            <property name="checkNamesForContextStrategyByTokenOrAncestorSet" value="RightCurlyCheck" />
```
Comment (baratali) [inline src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/RightCurly/context/defaultContextConfig.xml:12]: No difference with neverSuppressedChecks results. We should add violation which will be suppressed after this update. Please add another try-catch (in the same method in Test.java), which produces violation, but will be suppressed by patch-filter.

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/AvoidNestedBlocks/context/defaultContextConfig.xml [hunk 17]
```diff
@@ -9,7 +9,7 @@
         <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionJavaPatchFilter">
             <property name="file" value="${tp}/defaultContext.patch" />
             <property name="strategy" value="context" />
-            <property name="neverSuppressedChecks" value="AvoidNestedBlocksCheck" />
+            <property name="checkNamesForContextStrategyByTokenOrAncestorSet" value="AvoidNestedBlocksCheck" />
```
Comment (baratali) [inline src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/AvoidNestedBlocks/context/defaultContextConfig.xml:12]: No difference with neverSuppressedChecks results. We should add violation which will be suppressed after this update. Please add another method with a nested block (to Test.java), which produces a violation, but will be suppressed by patch-filter.

**Review by baratali — CHANGES_REQUESTED** [review]

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/DesignForExtension/context/defaultContextConfig.xml (hunk 13 above)
Comment (HuGanghui) [reply src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/DesignForExtension/context/defaultContextConfig.xml]: removed

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/FallThrough/case1/context/defaultContextConfig.xml (hunk 14 above)
Comment (HuGanghui) [reply src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/FallThrough/case1/context/defaultContextConfig.xml]: removed

File: src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/FallThrough/case2/context/defaultContextConfig.xml (hunk 15 above)
Comment (HuGanghui) [reply src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionjavapatchfilter/FallThrough/case2/context/defaultContextConfig.xml]: removed

**Comment by baratali** [comment]
@HuGanghui , I merged this pr. Please address https://github.com/checkstyle/patch-filters/pull/303#discussion_r477654781 and https://github.com/checkstyle/patch-filters/pull/303#discussion_r477661310, and prepare new pull request.
